feat: tenant tagging pipeline draft

botanical opened the pull request:
**Summary:**

https://github.com/NASA-IMPACT/veda-architecture/issues/677

## Changes

* Created a DAG to tag a list of collections with a tenant property
* Added a function to fix normalizing dates, which I noticed when I was testing updating collections with my DAG

## PR Checklist

- [ ] Unit tests
- [x] Ad-hoc testing - Deploy changes and test manually
- [ ] Integration tests
- [ ] Infrastructure changes - test using `terraform validate` and `terraform plan`

botanical committed:
feat: tenant tagging pipeline draft

botanical committed:
fix: update to take in a list of collections

botanical committed:
fix: update to handle collections, normalizing dates, and add logging

botanical commented on dags/veda_data_pipeline/utils/collection_generation.py:
```diff
@@ -99,16 +99,15 @@ def create_cog_collection(self, dataset: Dict[str, Any]) -> dict:
-                        for x in list(temporal_extent.values())
-                    ]
-                ]
+                "interval": [normalized_values]
```
@smohiudd do you know if this is where the existing date bug is?

smohiudd replied:
I think it may be in the ingest API

botanical committed:
fix: revert collection generation changes

botanical committed:
fix: update dag to take in catalog endpoint as alternative parameter

anayeaye commented on dags/veda_data_pipeline/veda_tenant_tagging_pipeline.py:
```diff
@@ -0,0 +1,297 @@
+        type=["null", "string"],
+        description="STAC catalog endpoint URL to fetch all collections from (optional if collections is provided)"
+    ),
+    "tenant": Param(default=None, type="string", description="Tenant ID to tag the collection with (will be set as eic-tenant property)"),
```
Because tenant filter field is configurable in the backend should we use two keys in this config `tenant_filter_field` and `tenant_filter_value`?
https://github.com/NASA-IMPACT/veda-backend/blob/dbd05924d7c04eeeec9ed88cc445a7b33a312261/stac_api/runtime/src/config.py#L103

anayeaye commented on dags/veda_data_pipeline/veda_tenant_tagging_pipeline.py:
```diff
@@ -0,0 +1,297 @@
+        type=["null", "array"],
+        description="List of collection IDs to tag (optional if catalog_endpoint is provided)"
+    ),
+    "catalog_endpoint": Param(
```
Discussion: I would almost rather have this DAG accept an array of collection ids so that we can use it on a subset of the collections in a catalog. I'm still reading through the methods below for adding a tag to all existing collections in a catalog to think it over.
cc: @botanical @smohiudd

botanical replied:
Okay, let me know! It takes in either a catalog endpoint or a list of collection IDs so I can just modify it to remove the catalog endpoint parameter @anayeaye

anayeaye replied:
Oh cool--I saw the catalog first. The collection ids make sense to me and let's chat about pros/cons on updating an entire catalog. I'll resolve this and we can chat in stand up

botanical replied:
Talked offline, we're going to probably remove the collections catalog endpoint parameter since it's an unlikely scenario where we want to set a whole catalog to a single tenant in multi-tenancy

botanical committed:
fix: update dag to take in tenant field parameter

botanical committed:
fix: remove catalog endpoint parameter

botanical committed:
fix: update dag to place tenant at top level

botanical committed:
fix: split list of collections, this assumes no quotations

botanical committed:
fix: get strings, then split

botanical committed:
fix: add error handling to ingestion step

botanical committed:
fix: update to sequentially process collections

botanical committed:
fix: use AirflowException

botanical committed:
fix: update to handle fetch error

botanical committed:
fix: add logging to see request size before sending to ingest

anayeaye commented on dags/veda_data_pipeline/veda_tenant_tagging_pipeline.py:
```diff
@@ -0,0 +1,354 @@
+
+        updated_collection = existing_collection.copy()
+
+        if "properties" not in updated_collection:
```
Properties are not part of the spec for collections so we shouldn't require. I'm don't know that we ever would want to add a properties object to a collection. 🤔

botanical replied:
https://github.com/NASA-IMPACT/veda-data-airflow/pull/414/commits/1c384ca6d006c31cf9726a10f9d18e98f4da63d6

botanical committed:
fix: update to remove properties in tenant dag

sandrahoang686 commented on dags/veda_data_pipeline/utils/schemas.py:
```diff
@@ -13,3 +15,64 @@ class DatetimeInterval(TimeInterval):
+                    normalized_intervals.append(normalized_interval)
+                else:
+                    normalized_intervals.append(interval)
+            collection["extent"]["temporal"]["interval"] = normalized_intervals
```
⛏️ 
```suggestion
            collection["extent"]["temporal"]["interval"] = [
                [normalize_datetime_to_iso8601(dt) for dt in interval]
                if isinstance(interval, list)
                else interval
                for interval in temporal["interval"]
            ]
```
? maybe

botanical replied:
We could do it your suggested way and use list comprehension! I think it's functionally the same

sandrahoang686 commented on dags/veda_data_pipeline/veda_tenant_tagging_pipeline.py:
```diff
@@ -0,0 +1,327 @@
+        if not isinstance(collections, list):
+            error_msg = f"Collections must be a list, but got type: {type(collections)}"
+            logger.error(error_msg)
+            raise ValueError(error_msg)
```
I'm just running through these cases and I was wondering - so are we expecting at line 86 to get a string like so `"col1, col2, col3"`? If so and line 87 proceeds sucessfully then does condition 92 and 97 ever get reached? should 92 be checking for `collections_strings` instead? and then 97 might not be needed because it wont reach - if `split()` which should always return an array doesn't succeed, it errors out right away and proceeds to the exception I think? but not sure

botanical replied:
Thanks for all the questions @sandrahoang686! I think you're right that line 997 might not be reachable 👀 since line 87, with `split()` will return a list. Let me simplify this so that the pipeline doesn't try to validate and parse a list of lists as input, and instead only accepts a list of strings

botanical replied:
https://github.com/NASA-IMPACT/veda-data-airflow/pull/414/changes/80f75d62fb57314bfe4b3249c033f7d6e9a80a76

botanical committed:
fix: update get_collection_ids to handle str lists only, and error otherwise

botanical committed:
Update dags/veda_data_pipeline/utils/schemas.py

Co-authored-by: sandrahoang686 <sandrahoang686@gmail.com>

botanical committed:
fix: update and handle case with newlines, according to convention from airflow docs

botanical commented:
@sandrahoang686 Thanks for the review, I believe I addressed all your feedback. 

I made one additional change to update the DAG to handle newlines for a collections list since this seems to be the convention that airflow uses for list params. 
https://airflow.apache.org/docs/apache-airflow/stable/_modules/airflow/example_dags/example_params_ui_tutorial.html
https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/params.html#use-params-to-provide-a-trigger-ui-form

ividito commented on dags/veda_data_pipeline/veda_tenant_tagging_pipeline.py:
```diff
@@ -0,0 +1,348 @@
+    update_collections = update_collection_with_tenant_tags.expand(existing_collection=fetch_collections)
+    ingest_collections = ingest_all_collections(collections=update_collections)
+
+    start >> collection_ids >> fetch_collections >> update_collections >> ingest_collections >> end
```
Style change to only define DAG edges in one place (just a preference)

```suggestion
    collection_ids = start >> get_collection_ids()
    fetch_collections = fetch_existing_collection.expand(collection_id=collection_ids)
    update_collections = update_collection_with_tenant_tags.expand(existing_collection=fetch_collections)
    ingest_collections = ingest_all_collections(collections=update_collections) >> end
```

ividito commented on dags/veda_data_pipeline/veda_tenant_tagging_pipeline.py:
```diff
@@ -0,0 +1,348 @@
+template_dag_run_conf = {
+    "collections": Param(
+        default=None,
+        type=["null", "array"],
```
Is "null" ever applicable here? If we have this as just "array" (making it required), does it still work to only input one collection ID in the UI? Apologies if this is a duplicate of @sandrahoang686's earlier questions

botanical replied:
Ah thanks for catching that! Yeah, null is not applicable here. It does still work if we input one collection ID but it needs to be an array (so a single element in an array).

botanical replied:
https://github.com/NASA-IMPACT/veda-data-airflow/pull/414/commits/b24a824bd67a5f87e743e0c09824a9db6b231516

ividito commented on dags/veda_data_pipeline/veda_tenant_tagging_pipeline.py:
```diff
@@ -0,0 +1,348 @@
+            time.sleep(0.5)
+
+    successful = sum(1 for r in results if r.get("status") == "success")
+    failed = sum(1 for r in results if r.get("status") == "error")
```
nit (`True` evaluates to 1 for `sum`)
```suggestion
    successful = sum(r.get("status") == "success" for r in results)
    failed = sum(r.get("status") == "error" for r in results)
```

ividito reviewed (approved):
Some nits but no blockers

botanical committed:
Update dags/veda_data_pipeline/veda_tenant_tagging_pipeline.py

Co-authored-by: Isayah Vidito <isayah@developmentseed.org>

botanical committed:
fix: update template daf run conf for collections to make required, array

botanical committed:
Merge branch 'issue-677/tenant-tagging-dag' of https://github.com/NASA-IMPACT/veda-data-airflow into issue-677/tenant-tagging-dag

botanical committed:
Update dags/veda_data_pipeline/veda_tenant_tagging_pipeline.py

Co-authored-by: Isayah Vidito <isayah@developmentseed.org>
